himembind.cpp

```diff
@@ -16,71 +16,126 @@
 #     include <process.h>
 #     define execvp _execvp
 #else
+
 #     include <unistd.h>
+
 #endif  // _MSC_VER
-#include <hwloc.h>
-#include <stdio.h>
-#include <stdlib.h>
 
+#include <hwloc.h>
 #include <memory>
+#include <algorithm>
+#include <cctype>
+#include <string_view>
+#include <iostream>
+
+bool is_valid_call_name(const char *call_name);
+
+bool is_valid_arguments(char **args, int argc);
+
+int main(int argc, char **argv) {
+    int ret = 0;
+
+    const char *myname = argv[0];
+    ++argv;
+    --argc;
+
+    const char *callname = argv[0];
+
+    hwloc_topology_t topology;
+    unsigned long flags = 0;
+
+    hwloc_topology_init(&topology);
+    hwloc_topology_set_flags(topology, flags);
+    hwloc_topology_load(topology);
+
+    hwloc_cpuset_t cpuset = hwloc_bitmap_alloc();
+    hwloc_get_cpubind(topology, cpuset, HWLOC_CPUBIND_PROCESS);
+
+    char *ss = nullptr;
+    hwloc_bitmap_asprintf(&ss, cpuset);
+    std::cerr << "cpuset: " << ss << std::endl;
+    free(ss);
+    ss = nullptr;
+
+    char *ss1 = nullptr;
+    hwloc_nodeset_t nodeset = hwloc_bitmap_alloc();
+    hwloc_cpuset_to_nodeset(topology, cpuset, nodeset);
+    hwloc_bitmap_asprintf(&ss1, nodeset);
+    std::cerr << "nodeset: " << ss1 << std::endl;
+
+    free(ss1);
+    ss1 = nullptr;
+
+    ret = hwloc_set_membind_nodeset(topology, nodeset, HWLOC_MEMBIND_BIND,
+                                    HWLOC_MEMBIND_STRICT);
+    if (ret) {
+        std::cerr << myname << ": Failed to set membind policy" << std::endl;
+        goto exit_failure;
+    }
+
+    if (argc == 0) {
+        std::cerr << "Usage: " << myname << " <cmd>" << std::endl;
+        goto exit_failure;
+    }
+
+    // Check for call name
+    if (!is_valid_call_name(callname)) {
+        std::cerr << "Invalid call name" << std::endl;
+        goto exit_failure;
+    }
+
+    // Check for arguments
+    if (!is_valid_arguments(argv, argc)) {
+        std::cerr << "Invalid arguments" << std::endl;
+        goto exit_failure;
+    }
```

Comment:
`std::cerr <<` operator can throw an exception and then `goto exit_failure` will not be executed. Under the `exit_failure` label there is a code which deallocates data structures from memory and so memory will not be freed correctly. I think for security reasons it is prohibited to have such behavior.

Reply:
According to the cpp standard, it seems that the function of std::cerr is redirect output stream into stderr (http://www.cplusplus.com/reference/iostream/cerr/). Is there anything I missed that would throw an exception when I use std:cerr?

Reply:
I am not a fan of `goto` in the first place... and if it is just there to ensure that proper deallocation occurs, then we should use RAII instead.
I will commit a patch to address that.

Reply:
I have now pushed my changes. Using RAII (via a scope_guard, which is license-free BTW), we can get rid of the `goto`, use `std:cerr` and guarantee that deallocation will always take place.
When compiling with C++17, we even have the guarantee that the deallocation functions will be `noexcept`.